components/parquet_ext/src/async_arrow_writer.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+//! Async arrow writer for parquet file.
+
+use std::{
+    io::Write,
+    sync::{Arc, Mutex},
+};
+
+use arrow::{datatypes::SchemaRef, record_batch::RecordBatch};
+use parquet::{
+    arrow::ArrowWriter,
+    errors::{ParquetError, Result},
+    file::properties::WriterProperties,
+    format::{FileMetaData, KeyValue},
+};
+use tokio::io::{AsyncWrite, AsyncWriteExt};
+
+#[derive(Clone, Default)]
+pub struct SharedBuffer {
+    /// The buffer to store the data to be written.
+    ///
+    /// The lock is used to obtain internal mutability, so no worry about the
+    /// lock contention.
+    buffer: Arc<Mutex<Vec<u8>>>,
+}
+
+impl Write for SharedBuffer {
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        let mut buffer = self.buffer.lock().unwrap();
+        Write::write(&mut *buffer, buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        let mut buffer = self.buffer.lock().unwrap();
+        Write::flush(&mut *buffer)
+    }
+}
+
+/// Async arrow writer for parquet file.
+///
+/// A shared buffer is provided to the sync writer [ArrowWriter] and it will
+/// accept the data from the sync writer and flush the received data to the
+/// async writer when the buffer size exceeds the threshold.
+pub struct AsyncArrowWriter<W> {
+    sync_writer: ArrowWriter<SharedBuffer>,
+    async_writer: W,
+    shared_buffer: SharedBuffer,
+    max_buffer_size: usize,
+}
+
+impl<W: AsyncWrite + Unpin + Send> AsyncArrowWriter<W> {
+    pub fn try_new(
+        writer: W,
+        arrow_schema: SchemaRef,
+        max_buffer_size: usize,
+        props: Option<WriterProperties>,
+    ) -> Result<Self> {
+        let shared_buffer = SharedBuffer::default();
```

Comment:
`SharedBuffer::with_capacity(max_buffer_size)`

Reply:
Maybe we should let it grow by itself, whose cost will be amortized by the whole encoding procedure, otherwise it may consume much memory if a very large `max_buffer_size` is provided for an unlimited purpose.